godot-macros/src/class/derive_godot_class.rs

```diff
@@ -62,8 +64,8 @@ pub fn derive_godot_class(item: venial::Item) -> ParseResult<TokenStream> {
     let godot_exports_impl = make_property_impl(class_name, &fields);
 
     let godot_withbase_impl = if let Some(Field { name, .. }) = &fields.base_field {
-        quote! {
-            impl ::godot::obj::WithBaseField for #class_name {
+        let implementation = if fields.well_formed_base {
```

Comment:
Maybe add a comment that we still need to generate code in case of bad base syntax, so that spans are retained/propagated to the compiler error.

Reply:
though this does actually cause an issue here:
![Screenshot_20241016_191646](https://github.com/user-attachments/assets/628edb65-59f3-4434-85ff-dcfc0cc0dd13)
where there is only an error at the `#[var]`, and no error corresponding to the type being wrong like here:
![Screenshot_20241016_191714](https://github.com/user-attachments/assets/7d30231a-ff88-455a-8e24-cca947312127)
i wonder if there's a good way to fix both

Reply:
i've mostly figured out a nicer way to fix this actually, but one problem i keep having is that i can't find a good way to get rust to report a nice error when the type is wrong. or rather to stop reporting a bad error. because the init auto generated code creates code like this (simplified some long paths for readability):
```rs
impl GodotDefault for Foo {
    fn __godot_user_init(base: Base<Self::Base>) -> Self {
        Self {
            base: base,
        }
    }
}
```  

and if the user declares the base field to have the type say `Option<Gd<Node2D>>`, then rust will report an error that an `Option<Gd<Node2D>>` was expected but got a `Base<Node2D>`. which is kinda backwards, we want rust to report that a `Base<Node2D>` was expected but got an `Option<Gd<Node2D>>`. we can relatively easily add an error message of that kind, but the original error remains anyway.

Reply:
actually i think there's a way to do it, it involves making a custom trait like this:
```rs
#[diagnostic::on_unimplemented(
    message = "expected base `{Self}` got `{A}`",
    label = "expected base `{Self}` got `{A}`"
)]
#[doc(hidden)]
pub trait IsBase<T: GodotClass, A> {
    fn conv(b: Base<T>) -> A;
}
impl<T: GodotClass> IsBase<T, Base<T>> for Base<T> {
    fn conv(b: Base<T>) -> Base<T> {
        b
    }
}
```
which is used like this in the codegen:
```rs
impl GodotDefault for Foo {
    fn __godot_user_init(
        base: Base<Node2D>,
    ) -> Self {
        Self {
            base: <Base<Node2D> as IsBase<Node2D, Option<Gd<Node2D>>>>::conv(base),
        }
    }
}
```

Reply:
Hm, nice idea! But I'm not really sure that the error message is very understandable. Is the benefit mostly pointing to the type instead of one of the proc-macro attributes?

We may also consider that:
- using `#[hint(base)]` is rare
- accidentally typing `base` but not `Base<T>` is also not a common error
- this extra trait may add indirections that have negative effects in other places (including error messages)

So maybe this PR could focus on the more general problem for now (also to not make the diff too complex)?

Reply:
> Hm, nice idea! But I'm not really sure that the error message is very understandable. Is the benefit mostly pointing to the type instead of one of the proc-macro attributes?

It's to have an error saying that a `Base<_>` was expected but an `Option<Gd<T>>` was given, instead of an error saying `Option<Gd<_>>` was expected but a `Base<_>` was given

> We may also consider that:
> 
>     * using `#[hint(base)]` is rare
> 
>     * accidentally typing `base` but not `Base<T>` is also not a common error
> 
>     * this extra trait may add indirections that have negative effects in other places (including error messages)
> 
> 
> So maybe this PR could focus on the more general problem for now (also to not make the diff too complex)?

Reply:
I read the new error message again. The main points are quite helpful, but this one is confusing:
> help: the trait `IsBase<godot::prelude::Node2D, Option<godot::prelude::Gd<godot::prelude::Node2D>>>` is not implemented for `Base<godot::prelude::Node2D>`

People would then look for an `IsBase` trait (it becomes a leaky abstraction). Unfortunately, it seems like there's only so much we can do regarding good error messages. In this case, I hope the error occurs rare enough that we don't need to add machinery to make it slightly better.

So you think you can revert the `IsBase` addition?